Quick stats (#21436)
prestodb/presto: presto-docs/src/main/sphinx/connector/hive.rst

* ``partition1_value1, partition1_value2``
* ``partition2_value1, partition2_value2``

Quick Stats
--------------------------------------

The Hive connector can build basic statistics for partitions with missing statistics

> by examining file or table metadata. For example, parquet footers can be used to infer

Reviewer: ```suggestion
by examining file or table metadata. For example, Parquet footers can be used to infer
```

> row counts, number of null and min/max values. These statistics help in query planning,

Reviewer: ```suggestion
row counts, number of nulls, and min/max values. These statistics help in query planning,
```
I'm incorporating your typo correction into my suggestion, which by itself I'm merely suggesting adding a comma.

> and serve as interim cache of stats for partitions which haven't yet been ANALYZEd

Reviewer: ```suggestion
and serve as an interim cache of stats for partitions which haven't yet been ANALYZEd. 
```
Minor grammar and punctuation. ANALYZEd feels a little odd - not wrong, but I'm not sure when I read this what it means. Do you have another way that you could describe what you mean here?

Author: I've rewritten this to `These 'quick' statistics help in query planning,
and serve as as a temporary source of stats for partitions which haven't had ANALYZE run on
them.`

The following properties can be used to control how these quick stats are built:

.. list-table::
   :widths: 20 70 10
   :header-rows: 1

   -

      - Property Name
      - Description
      - Default
   -

      - ``hive.quick-stats.enabled``
      - Enable stats collection through quick stats providers. Also

>          toggleable through the ``quick_stats_enabled`` session property

Reviewer: ```suggestion
        toggleable through the ``quick_stats_enabled`` session property.
```
Every text line after the first in the Description column has one extra space indent at the beginning. I'll try to edit these here, but if you can do a local build of the docs it'd help. I'm including screenshot of the build of the current file, and a screenshot of my local edited build to show the difference. I'm sorry this is going to be a lot of comments for almost entirely single-space deleting nits! 
current
<img width="772" alt="Screenshot 2024-02-07 at 12 24 08 PM" src="https://github.com/prestodb/presto/assets/7013443/81a9ae49-212e-493e-840d-48e2406f1215">

my edited and local-built example
<img width="760" alt="Screenshot 2024-02-07 at 12 33 55 PM" src="https://github.com/prestodb/presto/assets/7013443/0d51df58-61b7-4cab-b700-78eb0eb4123d">

Author: My bad, I used pandoc to convert a markdown table to rst and missed this. Will fix this up

      - ``false``
   -

      - ``hive.quick-stats.max-concurrent-calls``
      - Quick stats are built for multiple partitions concurrently. This

>          property sets the maximum number of concurrent builds that can

Reviewer: ```suggestion
        property sets the maximum number of concurrent builds that can
```

>          be made

Reviewer: ```suggestion
        be made.
```

      - ``100``
   -

      - ``hive.quick-stats.inline-build-timeout``
      - Duration the query that initiates a quick stats build for a

>          partition should wait before timing out and returning empty

Reviewer: ```suggestion
        partition should wait before timing out and returning empty
```

>          stats. Also toggleable through the

Reviewer: ```suggestion
        stats. Also toggleable through the
```

>          ``quick_stats_inline_build_timeout`` session property

Reviewer: ```suggestion
        ``quick_stats_inline_build_timeout`` session property.
```

      - ``60s``
   -

      - ``hive.quick-stats.background-build-timeout``

>       - If a query observes that quick stats are already being built for

Reviewer: ```suggestion
      - If a query observes that quick stats are being built for
```

>          a partition, this is the duration it will wait for those stats

Reviewer: ```suggestion
        a partition, this is the duration it waits for those stats
```

>          to be built before returning empty stats

Reviewer: ```suggestion
        to be built before returning empty stats.
```

      - ``0s``
   -

      - ``hive.quick-stats.cache-expiry``

>       - Duration to retain the stats in the quick stats in-memory cache

Reviewer: ```suggestion
      - Duration to retain the stats in the quick stats in-memory cache.
```
Nit: If one item in a list needs punctuation (a period), they all do. (I wanted to explain why I made this and some of the similar nit suggestions.)

      - ``24h``
   -

      - ``hive.quick-stats.reaper-expiry``

>       - If the quick stats build for a partition is stuck (say due to

Reviewer: ```suggestion
      - If the quick stats build for a partition is stuck (for example, due to
```

>          long-running IO), a reaper thread will terminate any background

Reviewer: ```suggestion
        long-running IO), a reaper thread terminates any background
```

>          build threads so that a new fetch could be triggered afresh.

Reviewer: ```suggestion
        build threads so that a new fetch could be triggered afresh.
```

>          This property controls the duration, after a background build

Reviewer: ```suggestion
        This property controls the duration, after a background build
```

>          thread is started, for the reaper to perform the termination

Reviewer: ```suggestion
        thread is started, for the reaper to perform the termination.
```

      - ``5m``
   -

      - ``hive.quick-stats.parquet.max-concurrent-calls``
      - Multiple parquet file footers are read and processed

>          concurrently. This property sets the maximum number of such

Reviewer: ```suggestion
        concurrently. This property sets the maximum number of
```

>          concurrent calls that can be made

Reviewer: ```suggestion
        concurrent calls that can be made.
```

      - ``500``
   -

      - ``hive.quick-stats.parquet.file-metadata-fetch-timeout``
      - Duration after which the parquet quick stats builder will fail

>          and return empty stats.

Reviewer: ```suggestion
        and return empty stats.
```

        Also toggleable through the ``quick_stats_inline_build_timeout``
        session property.
      - ``60s``
   -

      - ``hive.quick-stats.background-build-timeout``
      - If a query observes that quick stats are being built for
        a partition by another query, this is the duration it waits for
        those stats to be built before returning empty stats.
        Set this to ``0s`` if you want only one query to wait for
        quick stats to be built (for a given partition).
      - ``0s``
   -

      - ``hive.quick-stats.cache-expiry``
      - Duration to retain the stats in the quick stats in-memory cache.
      - ``24h``
   -

      - ``hive.quick-stats.reaper-expiry``
      - If the quick stats build for a partition is stuck (for example, due to
        a long-running IO operation), a reaper job terminates any background
        build threads so that a new fetch could be triggered afresh.
        This property controls the duration, after a background build
        thread is started, for the reaper to perform the termination.
      - ``5m``
   -

      - ``hive.quick-stats.parquet.max-concurrent-calls``

>       - Multiple parquet file footers are read and processed

Reviewer: ```suggestion
      - Multiple Parquet file footers are read and processed
```

        concurrently. This property sets the maximum number of
        concurrent calls that can be made.
      - 500
   -

      - ``hive.quick-stats.parquet.file-metadata-fetch-timeout``

>       - Duration after which the parquet quick stats builder will fail

Reviewer: ```suggestion
      - Duration after which the Parquet quick stats builder will fail
```
